{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-06-15 10:09:51,048 - (<class 'ZeroDivisionError'>, ZeroDivisionError('division by zero'), <traceback object at 0x000001AC0E409540>)\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-4-5ddf863d0c63>\", line 19, in <module>\n",
      "    print(1/0)\n",
      "ZeroDivisionError: division by zero\n",
      "2021-06-15 10:09:51,048 - (<class 'ZeroDivisionError'>, ZeroDivisionError('division by zero'), <traceback object at 0x000001AC0E409540>)\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-4-5ddf863d0c63>\", line 19, in <module>\n",
      "    print(1/0)\n",
      "ZeroDivisionError: division by zero\n",
      "2021-06-15 10:09:51,048 - (<class 'ZeroDivisionError'>, ZeroDivisionError('division by zero'), <traceback object at 0x000001AC0E409540>)\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-4-5ddf863d0c63>\", line 19, in <module>\n",
      "    print(1/0)\n",
      "ZeroDivisionError: division by zero\n",
      "2021-06-15 10:09:51,048 - (<class 'ZeroDivisionError'>, ZeroDivisionError('division by zero'), <traceback object at 0x000001AC0E409540>)\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-4-5ddf863d0c63>\", line 19, in <module>\n",
      "    print(1/0)\n",
      "ZeroDivisionError: division by zero\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-4-5ddf863d0c63>\", line 19, in <module>\n",
      "    print(1/0)\n",
      "ZeroDivisionError: division by zero\n"
     ]
    }
   ],
   "source": [
    "# _*_ logging: try/exception practices_*_ #\n",
    "\n",
    "# _*_ traceback.print_exc(file=open(filename, mode='a', encoding='utf-8')) _*_ #\n",
    "\n",
    "import logging\n",
    "import sys\n",
    "import traceback\n",
    "def get_logger():\n",
    "    logger = logging.getLogger()\n",
    "    logger.setLevel(logging.INFO)\n",
    "    \n",
    "    handler = logging.StreamHandler()\n",
    "    formatter = logging.Formatter('%(asctime)s - %(message)s')\n",
    "    \n",
    "    handler.setFormatter(formatter)\n",
    "    logger.addHandler(handler)\n",
    "    \n",
    "    return logger\n",
    "\n",
    "logger = get_logger()\n",
    "try:\n",
    "    print(1/0)\n",
    "except Exception as e:\n",
    "    logger.exception(sys.exc_info())\n",
    "    #logger.info(e) / logger.debug(e)\n",
    "    traceback.print_exc()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Python所有标准异常类\n",
    "    BaseException\t所有异常的基类\n",
    "    SystemExit\t解释器请求退出\n",
    "    KeyboardInterrupt\t用户中断执行(通常是输入^C)\n",
    "    Exception\t常规错误的基类\n",
    "    StopIteration\t迭代器没有更多的值\n",
    "    GeneratorExit\t生成器(generator)发生异常来通知退出\n",
    "    SystemExit\tPython 解释器请求退出\n",
    "    StandardError\t所有的内建标准异常的基类\n",
    "    ArithmeticError\t所有数值计算错误的基类\n",
    "    FloatingPointError\t浮点计算错误\n",
    "    OverflowError\t数值运算超出最大限制\n",
    "    ZeroDivisionError\t除(或取模)零 (所有数据类型)\n",
    "    AssertionError\t断言语句失败\n",
    "    AttributeError\t对象没有这个属性\n",
    "    EOFError\t没有内建输入,到达EOF 标记\n",
    "    EnvironmentError\t操作系统错误的基类\n",
    "    IOError\t输入/输出操作失败\n",
    "    OSError\t操作系统错误\n",
    "    WindowsError\t系统调用失败\n",
    "    ImportError\t导入模块/对象失败\n",
    "    KeyboardInterrupt\t用户中断执行(通常是输入^C)\n",
    "    LookupError\t无效数据查询的基类\n",
    "    IndexError\t序列中没有没有此索引(index)\n",
    "    KeyError\t映射中没有这个键\n",
    "    MemoryError\t内存溢出错误(对于Python 解释器不是致命的)\n",
    "    NameError\t未声明/初始化对象 (没有属性)\n",
    "    UnboundLocalError\t访问未初始化的本地变量\n",
    "    ReferenceError\t弱引用(Weak reference)试图访问已经垃圾回收了的对象\n",
    "    RuntimeError\t一般的运行时错误\n",
    "    NotImplementedError\t尚未实现的方法\n",
    "    SyntaxError\tPython 语法错误\n",
    "    IndentationError\t缩进错误\n",
    "    TabError\tTab 和空格混用\n",
    "    SystemError\t一般的解释器系统错误\n",
    "    TypeError\t对类型无效的操作\n",
    "    ValueError\t传入无效的参数\n",
    "    UnicodeError\tUnicode 相关的错误\n",
    "    UnicodeDecodeError\tUnicode 解码时的错误\n",
    "    UnicodeEncodeError\tUnicode 编码时错误\n",
    "    UnicodeTranslateError\tUnicode 转换时错误\n",
    "    Warning\t警告的基类\n",
    "    DeprecationWarning\t关于被弃用的特征的警告\n",
    "    FutureWarning\t关于构造将来语义会有改变的警告\n",
    "    OverflowWarning\t旧的关于自动提升为长整型(long)的警告\n",
    "    PendingDeprecationWarning\t关于特性将会被废弃的警告\n",
    "    RuntimeWarning\t可疑的运行时行为(runtime behavior)的警告\n",
    "    SyntaxWarning\t可疑的语法的警告\n",
    "    UserWarning\t用户代码生成的警告\n",
    "'''\n",
    "# _*_ try/except processing _*_ #\n",
    "'''\n",
    "一个try，至少有一个except处理异常:\n",
    "1.指定异常类：\n",
    "    try: {}\n",
    "    except ErrorType1: {} #非指定异常，无法处理\n",
    "    except ErrorType2: {}\n",
    "    ...\n",
    "    \n",
    "2.万能异常：\n",
    "    try: {}\n",
    "    except Exception as e: {}\n",
    "    \n",
    "3.try ... finally\n",
    "    try:\n",
    "        action\n",
    "    except:\n",
    "        exception action\n",
    "    else:\n",
    "        无异常时，action\n",
    "    finally:\n",
    "        无论异常与否，都会执行\n",
    "\n",
    "4.主动抛出异常：raise Exception\n",
    "\n",
    "5.自定义异常：\n",
    "    class Networkerror(BaseException):\n",
    "        def __init__(self,msg):\n",
    "            self.msg=msg\n",
    "        def __str__(self):\n",
    "            return self.msg\n",
    "\n",
    "    try:\n",
    "        raise Networkerror('类型错误')\n",
    "    except Networkerror as e:\n",
    "        print(e)\n",
    "        \n",
    "6.断言assert\n",
    "\n",
    "'''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
